### PR TITLE
Fix compiler deadlock

### PR DESCRIPTION
**Pull request by BlakeWilliams**
Currently there's a deadlock in the compiler due to rendering locking,
which can lead to edgecases where other locks prevent the completion of
rendering and the inability to complete rendering prevents other locks
from releasing as seen in https://github.com/github/view_component/issues/1447

This resolves the issue by replacing `Monitor` with a `ReadWriteLock`.
This allows the same component to render without requiring an exclusive
lock, resulting in concurrent renders.

In summary:

- Compilation will hold a write lock, blocking `#compile` and
  `render_template_for` calls.
- `render_template_for` calls will hold a read lock, blocking `#compile`
  but allowing other calls to `render_template_for` to check out another
  read lock.

See the documentation for `ReadWriteLock` for more details: http://ruby-concurrency.github.io/concurrent-ruby/master/Concurrent/ReadWriteLock.html

closes https://github.com/github/view_component/issues/1447

### CODE REVIEW

File: test/sandbox/test/rendering_test.rb [hunk 1]
```diff
@@ -1078,4 +1078,30 @@ def test_renders_nested_collection
...
+        t = Thread.new do
+          render_inline(ContentEvalComponent.new) do
+            mutex.synchronize do
+              sleep 0.01
+            end
+          end
+        end
+
+        t1.join
+        t.join
```
Comment (BlakeWilliams) [inline test/sandbox/test/rendering_test.rb:1103]: There's no assertion here, since the old code causes an error that looks like this:

```
Error:
RenderingTest#test_concurrency_deadlock:
fatal: No live threads left. Deadlock?
7 threads, 7 sleeps current:0x000000010697e260 main thread:0x00000001516046f0
* #<Thread:0x000000015205fae8 sleep_forever>
   rb_thread_t:0x00000001516046f0 native:0x00000001031ac580 int:0
   /Users/blake/Code/view_component/test/sandbox/test/rendering_test.rb:1103:in `join'
```

followed by a lot of stacktrace information 😄

**Review by joelhawksley — APPROVED** [review]
